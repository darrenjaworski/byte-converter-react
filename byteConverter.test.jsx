import React from "react";
import renderer from "react-test-renderer";

import ByteConverter from "./index";

describe("ByteConverter", () => {
  it("renders without crashing", () => {
    const tree = renderer
      .create(<ByteConverter>{1024}</ByteConverter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("can convert bytes correctly", () => {
    const MBvalue = renderer
      .create(
        <ByteConverter inUnit="KB" outUnit="MB">
          1024
        </ByteConverter>
      )
      .toJSON();

    const GBvalue = renderer
      .create(
        <ByteConverter useSI inUnit="MB" outUnit="GB">
          1000
        </ByteConverter>
      )
      .toJSON();

    expect(MBvalue).toBe("1");
    expect(GBvalue).toBe("1");
  });

  // permutations
  const units = ["B", "KB", "MB", "GB", "TB", "PB"];

  units.forEach(InUnit => {
    units.forEach(OutUnit => {
      it(`inUnit ${InUnit}, outUnit ${OutUnit} matches snapshot`, () => {
        const tree = renderer
          .create(
            <ByteConverter inUnit={InUnit} outUnit={OutUnit}>
              {1024}
            </ByteConverter>
          )
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  it("should render with commas", () => {
    const commas = renderer
      .create(
        <ByteConverter addCommas inUnit="GB" outUnit="MB">
          {1}
        </ByteConverter>
      )
      .toJSON();
    expect(commas).toBe("1,024");
  });

  const unitSuffixes = {
    B: "byte",
    KB: "kilobyte",
    MB: "megabyte",
    GB: "gigabyte",
    TB: "terabyte",
    PB: "petabyte"
  };

  const suffixes = Object.keys(unitSuffixes);

  suffixes.forEach(suffix => {
    const suffixString = unitSuffixes[suffix];
    it(`should add the suffix ${suffixString}s`, () => {
      const result = renderer
        .create(
          <ByteConverter suffix inUnit="PB" outUnit={suffix}>
            {10}
          </ByteConverter>
        )
        .toJSON();

      const resultSuffix = result[1];
      expect(resultSuffix.trim()).toBe(`${suffixString}s`);
    });
  });

  it("should have a singular suffix when exactly one is resulting", () => {
    const result = renderer
      .create(
        <ByteConverter suffix inUnit="KB">
          {1024}
        </ByteConverter>
      )
      .toJSON();

    const resultSuffix = result[1];
    expect(resultSuffix.trim()).toBe("megabyte");
  });

  it("should warn with an infinite number", () => {
    const spyWarn = jest.spyOn(console, "warn");
    renderer.create(<ByteConverter>{-Infinity}</ByteConverter>);
    renderer.create(<ByteConverter>{Infinity}</ByteConverter>);
    expect(spyWarn).toHaveBeenCalledTimes(5);
  });

  it("should warn with a non integer number", () => {
    const spyWarn = jest.spyOn(console, "warn");
    renderer.create(<ByteConverter>{10.1}</ByteConverter>);
    expect(spyWarn).toHaveBeenCalledTimes(6);
  });

  it("should warn with a negative number", () => {
    const spyWarn = jest.spyOn(console, "warn");
    renderer.create(<ByteConverter>{-10}</ByteConverter>);
    expect(spyWarn).toHaveBeenCalledTimes(7);
  });

  it("should warn when not a number", () => {
    const spyWarn = jest.spyOn(console, "warn");
    renderer.create(<ByteConverter>blah</ByteConverter>);
    expect(spyWarn).toHaveBeenCalledTimes(11);
  });

  it("should suppress warnings when `hideWarnings` is present", () => {
    const spyWarn = jest.spyOn(console, "warn");
    renderer.create(<ByteConverter hideWarnings>{-Infinity}</ByteConverter>);
    renderer.create(<ByteConverter hideWarnings>{Infinity}</ByteConverter>);
    renderer.create(<ByteConverter hideWarnings>{10.1}</ByteConverter>);
    renderer.create(<ByteConverter hideWarnings>{-10}</ByteConverter>);
    expect(spyWarn).not.toHaveBeenCalledTimes(12);
  });
});
