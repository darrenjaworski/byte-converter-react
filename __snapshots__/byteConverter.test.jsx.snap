// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ByteConverter inUnit B, outUnit B matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit B, outUnit GB matches snapshot 1`] = `"9.5367431640625e-7"`;

exports[`ByteConverter inUnit B, outUnit KB matches snapshot 1`] = `"1"`;

exports[`ByteConverter inUnit B, outUnit MB matches snapshot 1`] = `"0.0009765625"`;

exports[`ByteConverter inUnit B, outUnit PB matches snapshot 1`] = `"9.094947017729282e-13"`;

exports[`ByteConverter inUnit B, outUnit TB matches snapshot 1`] = `"9.313225746154785e-10"`;

exports[`ByteConverter inUnit B, outUnit b matches snapshot 1`] = `"8192"`;

exports[`ByteConverter inUnit GB, outUnit B matches snapshot 1`] = `"1099511627776"`;

exports[`ByteConverter inUnit GB, outUnit GB matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit GB, outUnit KB matches snapshot 1`] = `"1073741824"`;

exports[`ByteConverter inUnit GB, outUnit MB matches snapshot 1`] = `"1048576"`;

exports[`ByteConverter inUnit GB, outUnit PB matches snapshot 1`] = `"0.0009765625"`;

exports[`ByteConverter inUnit GB, outUnit TB matches snapshot 1`] = `"1"`;

exports[`ByteConverter inUnit GB, outUnit b matches snapshot 1`] = `"8796093022208"`;

exports[`ByteConverter inUnit KB, outUnit B matches snapshot 1`] = `"1048576"`;

exports[`ByteConverter inUnit KB, outUnit GB matches snapshot 1`] = `"0.0009765625"`;

exports[`ByteConverter inUnit KB, outUnit KB matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit KB, outUnit MB matches snapshot 1`] = `"1"`;

exports[`ByteConverter inUnit KB, outUnit PB matches snapshot 1`] = `"9.313225746154785e-10"`;

exports[`ByteConverter inUnit KB, outUnit TB matches snapshot 1`] = `"9.5367431640625e-7"`;

exports[`ByteConverter inUnit KB, outUnit b matches snapshot 1`] = `"8388608"`;

exports[`ByteConverter inUnit MB, outUnit B matches snapshot 1`] = `"1073741824"`;

exports[`ByteConverter inUnit MB, outUnit GB matches snapshot 1`] = `"1"`;

exports[`ByteConverter inUnit MB, outUnit KB matches snapshot 1`] = `"1048576"`;

exports[`ByteConverter inUnit MB, outUnit MB matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit MB, outUnit PB matches snapshot 1`] = `"9.5367431640625e-7"`;

exports[`ByteConverter inUnit MB, outUnit TB matches snapshot 1`] = `"0.0009765625"`;

exports[`ByteConverter inUnit MB, outUnit b matches snapshot 1`] = `"8589934592"`;

exports[`ByteConverter inUnit PB, outUnit B matches snapshot 1`] = `"1152921504606847000"`;

exports[`ByteConverter inUnit PB, outUnit GB matches snapshot 1`] = `"1073741824"`;

exports[`ByteConverter inUnit PB, outUnit KB matches snapshot 1`] = `"1125899906842624"`;

exports[`ByteConverter inUnit PB, outUnit MB matches snapshot 1`] = `"1099511627776"`;

exports[`ByteConverter inUnit PB, outUnit PB matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit PB, outUnit TB matches snapshot 1`] = `"1048576"`;

exports[`ByteConverter inUnit PB, outUnit b matches snapshot 1`] = `"9223372036854776000"`;

exports[`ByteConverter inUnit TB, outUnit B matches snapshot 1`] = `"1125899906842624"`;

exports[`ByteConverter inUnit TB, outUnit GB matches snapshot 1`] = `"1048576"`;

exports[`ByteConverter inUnit TB, outUnit KB matches snapshot 1`] = `"1099511627776"`;

exports[`ByteConverter inUnit TB, outUnit MB matches snapshot 1`] = `"1073741824"`;

exports[`ByteConverter inUnit TB, outUnit PB matches snapshot 1`] = `"1"`;

exports[`ByteConverter inUnit TB, outUnit TB matches snapshot 1`] = `"1024"`;

exports[`ByteConverter inUnit TB, outUnit b matches snapshot 1`] = `"9007199254740992"`;

exports[`ByteConverter inUnit b, outUnit B matches snapshot 1`] = `"128"`;

exports[`ByteConverter inUnit b, outUnit GB matches snapshot 1`] = `"1.1920928955078125e-7"`;

exports[`ByteConverter inUnit b, outUnit KB matches snapshot 1`] = `"0.125"`;

exports[`ByteConverter inUnit b, outUnit MB matches snapshot 1`] = `"0.0001220703125"`;

exports[`ByteConverter inUnit b, outUnit PB matches snapshot 1`] = `"1.1368683772161603e-13"`;

exports[`ByteConverter inUnit b, outUnit TB matches snapshot 1`] = `"1.1641532182693481e-10"`;

exports[`ByteConverter inUnit b, outUnit b matches snapshot 1`] = `"1024"`;

exports[`ByteConverter renders without crashing 1`] = `"0.0009765625"`;
